docs/draft/17-contact-history.md

```diff
@@ -0,0 +1,133 @@
+
+### Modifications to Message Format
+
+The `ApplicationMetadataMessage` is modified to include 3 new Types:
+
+```
+message ApplicationMetadataMessage {
+  // Signature of the payload field
+  bytes signature = 1;
+  // This is the encoded protobuf of the application level message, i.e ChatMessage
+  bytes payload = 2;
+
+  // The type of protobuf message sent
+  Type type = 3;
+
+  enum Type {
+    // ...
+    HISTORY_CAPABILITY = 42;
+    HISTORY_REQUEST = 43;
+    HISTORY_REPLY = 44;
+  }
+}
+```
+
+## Advertising Capabilities
+
+- When a contact request is sent the sender MAY advertise its capabilities
+- When a contact request is accepted the receiver MAY advertises its own capabilities
+
+```
+HistoryCapability {
+  uint64 clock = 1;
+  string contactChatId = 2
+  map<string, int64> capability = 3;
+}
+```
+
+`HistoryCapability.capability` has the mapping between chatIds and time in seconds. This indicates the contact is willing to send the chat history for last X seconds for that particular chatId.
+
+```
+{
+  "0x123": -1, // all available chat history
+  "0x234": 3600, // last 1h
+  "0x345": 300, // last 5 minutes
+}
+```
+
+If no capabilities are advertised, it MUST be assumed that contact will not relay contact history.
+
+When a user changes its capabilities, a change signal MUST be sent to contacts.
+
+## Requesting Chat history
+
+The client chooses one of his online contacts at random that indicated capability to send the history for the target chat or community.
+
+**Request**
+
+A `ChatMessage` with the ContentType `HISTORY` and the timestamp of the last known message in the 1 on 1 chat.
+
+```
+HistoryRequest {
+  uint64 clock = 1;
+  string contactChatId = 2
+  uint64 requestedChatId = 3;
+  uint64 requestedLastTimestamp = 4;
+}
+```
+
+**Response**
+
+```
+HistoryResponse {
+  uint64 clock = 1;
+  string contactChatId = 2
+  uint64 requestedChatId = 3;
+  string message = 4;
+}
+```
+
+`HistoryResponse.message` contains the encoded original message.
+A `HistoryResponse` is sent for each message in the requested time range.
+
+## Requesting 1 on 1 chat history
+
+In this case, only 1 contact will be a match. If the contact has not advertised intention to send the chat history, no request should be sent.
+
+## Requesting a group chat history
+
+The client chooses one of the online contacts that are members in the group chat at random to request the group history and that has advertised intention to send the chat history of this group chat.
+If no history is received during time X, then the client can choose another contact to request the history
+
+## Requesting a public chat history
+
+The client chooses one of the online contacts that displayed intentions to send chat history for the public chat in question.
+
+
+## Requesting a community chat history
+
+The client chooses one of the online contacts that displayed intentions to send chat history for the community in question.
```

Comment:
In all of these scenarios, status nodes will have to store each message as waku message.
Something to take into consideration as this will have an impact on storage.

Also, it should be taking into account here in the spec that this in fact needs to happen (waku messages need to be stored separately).

Reply:
ah this ties into the fact that we should have a  `dont-store` flag on waku-2, I think there's should be an issue somewhere about it in the vac repos

Reply:
For reference and context, the Community History Archive Protocol requires the same https://github.com/status-im/specs/pull/164

Reply:
https://github.com/vacp2p/rfc/issues/441